### PR TITLE
Differentiable refactor

### PR DESCRIPTION
**Pull request by Huangzizhou**
Main changes:

1. Wrap `std::vector<std::shared_ptr<VariableToSimulation>>` into a class and hide for loops over the vector inside.
2. Remove `first_derivative_unweighted` in `AdjointForm`, it may cause NAN if the weight is zero in the old version.
3. Move some implementations into cpp files.
4. Use struct instead of json to store data in IntegrableFunctional.
5. Make enum class naming consistent.

### CODE REVIEW

**Review by arvigj — APPROVED** [review]

File: src/polyfem/solver/forms/adjoint_forms/ParametrizedProductForm.hpp [hunk 1]
```diff
@@ -9,16 +9,17 @@ namespace polyfem::solver
 	class ParametrizedProductForm : public ParametrizationForm
 	{
 	public:
-		ParametrizedProductForm(const CompositeParametrization &parametrizations) : ParametrizationForm(parametrizations)
+		ParametrizedProductForm(CompositeParametrization &&parametrizations) : ParametrizationForm(std::move(parametrizations))
```
Comment (teseoch) [inline src/polyfem/solver/forms/adjoint_forms/ParametrizedProductForm.hpp:12]: a bit hugly

File: src/polyfem/solver/forms/parametrization/Parametrization.hpp [hunk 2]
```diff
@@ -26,8 +26,8 @@ namespace polyfem::solver
 	class CompositeParametrization : public Parametrization
 	{
 	public:
-		CompositeParametrization() {}
-		CompositeParametrization(const std::vector<std::shared_ptr<Parametrization>> &parametrizations) : parametrizations_(parametrizations) {}
+		CompositeParametrization() : parametrizations_(std::vector<std::shared_ptr<Parametrization>>()) {}
+		CompositeParametrization(std::vector<std::shared_ptr<Parametrization>>&& parametrizations) : parametrizations_(parametrizations) {}
```
Comment (teseoch) [inline src/polyfem/solver/forms/parametrization/Parametrization.hpp:30]: here too

File: tests/test_diff.cpp [hunk 3]
```diff
@@ -155,7 +153,7 @@ TEST_CASE("laplacian", "[test_adjoint]")
 	Eigen::MatrixXd velocity_discrete;
 	sample_field(*states[0], velocity, velocity_discrete);
 
-	verify_adjoint(*nl_problem, x, velocity_discrete, 1e-7, 3e-5);
+	verify_adjoint(*nl_problem, x, velocity_discrete, 1e-7, 1e-8);
```
Comment (teseoch) [inline tests/test_diff.cpp:156]: nice! why the change?

**Review by teseoch — COMMENTED** [review]
you used var_ for func arguments, the conversion was to use var_ for fields.

don't change it :)

**Comment by Huangzizhou** [comment]
> you used var_ for func arguments, the conversion was to use var_ for fields.
> 
> don't change it :)

You mean in IntegrableFunctional? Fixed.

File: tests/test_diff.cpp (hunk 3 above)
Comment (Huangzizhou) [reply tests/test_diff.cpp:156]: there was a bug in laplacian, it's a miracle that it's been there for two years.
